src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1753 @@
+    }
+  }
+
+  private final Logger log = LoggerFactory.getLogger(GenericDatabaseDialect.class);
+  protected final AbstractConfig config;
+
+  /**
+   * Whether to map {@code NUMERIC} JDBC types by precision.
+   */
+  protected final boolean mapNumerics;
+  protected final String catalogPattern;
+  protected final String schemaPattern;
+  protected final Set<String> tableTypes;
+  protected final String jdbcUrl;
+  private final IdentifierRules defaultIdentifierRules;
+  private final AtomicReference<IdentifierRules> identifierRules = new AtomicReference<>();
+  private final Queue<Connection> connections = new ConcurrentLinkedQueue<>();
+  private volatile JdbcDriverInfo jdbcDriverInfo;
+
+  /**
+   * Create a new dialect instance with the given connector configuration.
+   *
+   * @param config the connector configuration; may not be null
+   */
+  public GenericDatabaseDialect(AbstractConfig config) {
+    this(config, IdentifierRules.DEFAULT);
+  }
+
+  /**
+   * Create a new dialect instance with the given connector configuration.
+   *
+   * @param config                 the connector configuration; may not be null
+   * @param defaultIdentifierRules the default rules for identifiers; may be null if the rules are
+   *                               to be determined from the database metadata
+   */
+  protected GenericDatabaseDialect(
+      AbstractConfig config,
+      IdentifierRules defaultIdentifierRules
+  ) {
+    this.config = config;
+    this.defaultIdentifierRules = defaultIdentifierRules;
+    this.jdbcUrl = config.getString(JdbcSourceConnectorConfig.CONNECTION_URL_CONFIG);
+    if (!(config instanceof JdbcSinkConfig)) {
+      mapNumerics = config.getBoolean(JdbcSourceTaskConfig.NUMERIC_PRECISION_MAPPING_CONFIG);
+      catalogPattern = config.getString(JdbcSourceTaskConfig.CATALOG_PATTERN_CONFIG);
+      schemaPattern = config.getString(JdbcSourceTaskConfig.SCHEMA_PATTERN_CONFIG);
+      tableTypes = new HashSet<>(config.getList(JdbcSourceTaskConfig.TABLE_TYPE_CONFIG));
+    } else {
+      mapNumerics = JdbcSourceTaskConfig.NUMERIC_PRECISION_MAPPING_DEFAULT;
+      catalogPattern = JdbcSourceTaskConfig.CATALOG_PATTERN_DEFAULT;
+      schemaPattern = JdbcSourceTaskConfig.SCHEMA_PATTERN_DEFAULT;
+      tableTypes = new HashSet<>(Arrays.asList(JdbcSourceTaskConfig.TABLE_TYPE_DEFAULT));
+    }
+  }
+
+  @Override
+  public Connection getConnection() throws SQLException {
+    // These config names are the same for both source and sink configs ...
+    String username = config.getString(JdbcSourceConnectorConfig.CONNECTION_USER_CONFIG);
+    Password dbPassword = config.getPassword(JdbcSourceConnectorConfig.CONNECTION_PASSWORD_CONFIG);
+    Properties properties = new Properties();
+    if (username != null) {
+      properties.setProperty("user", username);
+    }
+    if (dbPassword != null) {
+      properties.setProperty("password", dbPassword.value());
+    }
+    properties = addConnectionProperties(properties);
```

Comment:
The `addConnectionProperties(Properties)` method is an example of a protected method that allows subclasses to override which properties are used without having to duplicate all of the non-trivial logic in this method. Method like this make it easier to subclass without duplicating a lot of logic, but it makes the class bigger. IMO, small price to pay for smaller subclasses.